rust/lance/src/dataset/transaction.rs

```diff
@@ -245,22 +262,28 @@ impl Transaction {
                 }
                 _ => true,
             },
-            // Merge changes the schema, but preserves row ids, so the only operation
-            // it's compatible with is CreateIndex.
-            Operation::Merge { .. } => !matches!(&other.operation, Operation::CreateIndex { .. }),
+            // Merge changes the schema, but preserves row ids, so the only operations
+            // it's compatible with is CreateIndex and ReserveFragments.
+            Operation::Merge { .. } => !matches!(
+                &other.operation,
+                Operation::CreateIndex { .. } | Operation::ReserveFragments { .. }
+            ),
         }
     }
 
     fn fragments_with_ids<'a, T>(
         new_fragments: T,
         fragment_id: &'a mut u64,
+        reserved_fragment_ids: Range<u64>,
```

Comment:
This is for the path where we later use the reserved fragment ids in some kind of append/delete/rewrite operation.

E.g. let's assume we are performing some kind of task that will need two fragments.  Let's pretend the max fragment id is 5.  We issue a reserve fragments call.  The max fragment id is now 7.  The caller knows they have fragments 6 & 7.

Later, we decide to commit the operation.  By this time the max fragment id is 12.  We have to somehow indicate that we want to use fragments 6 & 7 and not generate new fragment ids in any call to `fragments_with_ids`.  We have a few options:

 * We could simply assume that any fragment id (in the transaction) that is not set to 0 must have been reserved earlier and keep it instead of generating a new id.  This would not require the user to provide `reserved_fragment_ids` but feels brittle.
 * We can ask the user to provide `reserved_fragment_ids` and always generate new ids when the id in the transaction is not in that range (what we do today).  This is slightly safer but still requires we trust the user not to be mistaken about what they've reserved.
 * We could be even more conservative.  We could collect a hashset of all ids.  Then, when we encounter a fragment with an id that is not 0 we can check the existing id set and, if the fragment id is not in use, we allow it.  If the fragment id is 0 we generate a new one and, if it isn't 0, we error out.

I went with the second option but now I think I might choose the third.